drivers/lora/sx126x.c

```diff
@@ -331,6 +331,11 @@ void SX126xSetOperatingMode(RadioOperatingModes_t mode)
 		sx126x_set_rx_enable(1);
 		break;
 
+	case MODE_SLEEP:
+		/* Additionally disable the DIO1 interrupt to save power */
```

Comment:
Is it common to other platforms (other than Nordic) that gpio interrupt consumes constant power?

Reply:
I can't speak for all platforms, but it wasn't something that needed to be managed on TI or Silicon Labs microcontrollers.
Are there platforms where the enabling/disabling of the interrupt is a problem?

Reply:
> Are there platforms where the enabling/disabling of the interrupt is a problem?

No, I am not aware of any. It just adds a bit overhead when entering/exiting sleep mode. I just haven't seen other drivers doing similar stuff, so I was wondering whether this enable/disable interrupts is just a workaround for single platform.

On the other hand the LoRa radio will rarely (no more than once a second) switch between modes, so the "overhead" isn't significant.

Reply:
> I just haven't seen other drivers doing similar stuff, so I was wondering whether this enable/disable interrupts is just a workaround for single platform.

Because my hardware designs are Nordic based, all of my out-of-tree drivers with interrupts do this in some form or another. If we're stretching, this also helps in the case where you have N available interrupt lines, N+1 GPIO's that want to be interrupts at some point, and only some subset of drivers enabled at once.

The statement regarding low overhead applies to all the other drivers as well, you generally only care about the idle case. Saving 15uA when your base consumption is 5uA is huge. Saving 15uA when the sensor is using 2mA is irrelevant.

Reply:
>Because my hardware designs are Nordic based, all of my out-of-tree drivers with interrupts do this in some form or another.

AFAIK, not all Nordic MCUs have this issue. For example nRF52832 is affected and nRF52840 is not. My only concern here is that enabling/disabling interrupts in runtime is needed for small part of MCUs. My approach with reducing power was to use gpio level interrupts, so that the "faulty" (according to errata) GPIOTE mechanism is not used. Last week I've developed #31823, so you can take a look there.

>If we're stretching, this also helps in the case where you have N available interrupt lines, N+1 GPIO's that want to be interrupts at some point, and only some subset of drivers enabled at once.

Makes sense.

>The statement regarding low overhead applies to all the other drivers as well, you generally only care about the idle case. Saving 15uA when your base consumption is 5uA is huge. Saving 15uA when the sensor is using 2mA is irrelevant.

I was referring to "time/processing overhead" when basically before/after each communication over LoRa you need to setup and cleanup GPIO interrupts. The "power overhead" that you describe is huge, I agree.

Reply:
> nRF52840 is not

It definitely is, you can see this in section 5.2.1.1 of the datasheet.
```
System ON, full 256 kB RAM retention, wake on any event: 2.35uA
System ON, full 256 kB RAM retention, wake on GPIOTE input: 17.37uA
```

>I was referring to "time/processing overhead"

So was I, just from the opposite angle. My point was that you are only transitioning between these modes rarely, so spending a few clock cycles to save significant power is worth it.

Reply:
Re #31823, I can see that being an alternative. You are trading cycles at setup/teardown for cycles during every interrupt however, which is not always a good tradeoff.
The question IMO is whether drivers should be cleaning up resources that they are not currently using.

Reply:
Thanks for pointing out that nrf52840 is affected as well. Seems like this issue got upgraded from bug to a feature. Quite confusing, because I was only looking at errata and didn't find it there.

>Re #31823, I can see that being an alternative. You are trading cycles at setup/teardown for cycles during every interrupt however, which is not always a good tradeoff.

If that would work (not sure yet, still under review), then the major advantage would be one place where that would be handled, instead of modifying many drivers that depend on the edge interrupt feature.

Thanks for all explanation in this thread and also for patches!

Reply:
> Seems like this issue got upgraded from bug to a feature

It feels like the GPIOTE using power when enabled is part of the hardware design (otherwise why would you have the PORT events). If its part of the design, then it's not an errata.

> Thanks for all explanation in this thread and also for patches!

No problem, the fewer commits I have to keep reapplying on my fork each release the better!